### PR TITLE
Implement DeriveSha change logic at Kore HardFork

### PR DESCRIPTION
**Pull request by hqjang-pepper**
## Proposed changes
This PR implements DeriveSha change logic at `Kore` HardFork.

Main changes are as follow:
- move DeriveShaOriginal's package from `statedb` to `types`
- Set derivesha implementation to automatically change from concat to original upon Kore HardFork.

Plus, I conducted these tests:
- CN can generate block & process transactions,receipts well before/after hardfork
- Whether EN syncs well before & after CN's hardfork 
-  Check if deriveSha is applied as a hard forked version even after the node is turned off and on.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -266,6 +263,9 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
 		return nil, err
 	}
 
+	// Initialize DeriveSha implementation
```
Comment (kjhman21) [inline blockchain/blockchain.go:266]: @hqjang-pepper 
What is the reason of this change? It seems the value is not changed.

File: node/cn/backend.go [hunk 2]
```diff
@@ -272,6 +272,10 @@ func New(ctx *node.ServiceContext, config *Config) (*CN, error) {
 		return nil, err
 	}
 	bc.SetCanonicalBlock(config.StartBlockNumber)
+	if chainConfig.IsKoreForkEnabled(bc.CurrentBlock().Number()) {
```
[the remark this answers is not in the document]
Comment (ian0371) [reply node/cn/backend.go:275]: @hyunsooda We cannot modify the genesis parameters in the production environment. (see https://github.com/klaytn/klaytn/blob/81db7606f5252d8fa9d9cc0ad4769a0ee182220b/blockchain/genesis.go#L270).

I agree removing duplicate deriveShaImpl initialization. `cn.New` is a great location for chain config modification:
https://github.com/klaytn/klaytn/blob/81db7606f5252d8fa9d9cc0ad4769a0ee182220b/node/cn/backend.go#L222

[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply node/cn/backend.go:275]: Test passed when I moved `IsKoreForkEnabled` function as you mentioned.
Let's have a future conversation where to put this logic. +`homi` issue

[the remark this answers is not in the document]
Comment (hyunsooda) [reply node/cn/backend.go:275]: Hi @ian0371,

1. I meant that we may make a change so that the `genesis.json` output of homi will write zero value to `deriveShaImpl` field by default. 
2. `if chainConfig.IsKoreForkEnabled(bc.CurrentBlock().Number()) {` This dependency requires setup blockchain instance before chain configuration init. Its instantiation is done at `NewBlockChain()` and I suggested inserting inside of it (near `bc.loadLaststate()` call). 

Let me know if I’m missing something.

[the remark this answers is not in the document]
Comment (ian0371) [reply node/cn/backend.go:275]: @hyunsooda 
1. `homi setup --cn-num 1` will already generate `genesis.json` with `deriveShaImpl=0`. If you meant `homi setup --cypress` should also generate `genesis.json` with `deriveShaImpl=0`, I'm not sure if we should do that because if it's used on the production environment, `kcn init` will fail due to the conflict with the existing `deriveShaImpl=2` from database.
2. You're right, we need an initialized blockchain instance. Your suggested location seems fine.

[the remark this answers is not in the document]
Comment (hyunsooda) [reply node/cn/backend.go:275]: @ian0371,

1. I found another reason that we should not modify the target of cypress `genesis.json` because the newbie is not able to sync due to the failure of verification by unmatched tx root. Because the very initial blocks include tx roots calculated with `concat` method.

This is just to clarify your explanation, Is there a condition that asserts `deriveShaImpl` method must not be modified when the `genesis.json` is initialized again? I couldn't find where it is and checked out that reinitialization was successfully done and ran.

Thanks.

[the remark this answers is not in the document]
Comment (ian0371) [reply node/cn/backend.go:275]: @hyunsooda 
I tested re-initialization with different deriveShaImpl and failed with the following error.
```
$ homi setup local --cypress --cn-num 1
$ bin/kcn --datadir ./data init ./data/genesis.json
... success
$ # modify genesis.json: deriveShaImpl = 2 -> 0
$ bin/kcn --datadir ./data init ./data/genesis.json
...
CRIT[11/02,13:32:12 +09] [19] Failed to write genesis block             err="database already contains an incompatible genesis block (have 9c8c16594f977841, new 1a67dab39e7e7afd)"
```

The checking logic is here: 
https://github.com/klaytn/klaytn/blob/507698b99c32551721281f85e5be1a629610f8f3/blockchain/genesis.go#L227-L230

How did you test to make it not fail?

[the remark this answers is not in the document]
Comment (hyunsooda) [reply node/cn/backend.go:275]: I've used a custom script that removes data dir and runs `init` command for easier testing and used it when I tested. Sorry for the noise and thanks for the ref.

[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply node/cn/backend.go:275]: @ian0371 maybe you didn't removed chaindata after you made a genesis block.

[the remark this answers is not in the document]
Comment (ian0371) [reply node/cn/backend.go:275]: @hqjang-pepper Yep, the failure is intended. It's to show that we need to be careful when changing the default value of `genesis.json` generated from homi.

**Comment by blukat29** [comment]
I will create another PR that succeeds this PR.
The new PR will allow dynamically changing the DeriveShaImpl with a new `governance.deriveshaimpl` governance item.

**Comment by blukat29** [comment]
Closing in favor of #1708.
